Introduce "cilium-health", a new tool for investigating cluster connectivity issues.

joestringer opened the pull request:
This series extends the Cilium API to provide access to known nodes in the cluster and their IPs via /healthz (The same as #2139), and adds a new program to the tree: `cilium-health`.

`cilium-health` can be run as a daemon, which will periodically probe other nodes in the cluster for connectivity, starting with basic ping (ICMP) access and HTTP probes to other nodes. By default, when `cilium` is run, it will launch `cilium-health` locally. The results of the probes will be cached and exposed over a REST API. `cilium-health` can also be run as a commandline client for the daemon, to fetch the connectivity status and format it. Currently, `cilium-health` exposes:

* `GET /hello`
    * Returns success code with empty body, used for probing 
*  `GET /healthz`
    * Daemon uptime
    * Node load (1min/5min/15min)
    * Cilium /healthz
* `GET /status`
    * Timestamp of probe
    * Local node name
    * List of nodes
        * Name
        * IP
        * Latency to IPs
        * Errors, if any, attempting to reach IPs
* `PUT /status/probe`
    * Similar to GET /status - same result, but it triggers a probe and only returns when the probe completes.

Sample output
```
$ cilium-health status
Probe time:   2017-11-27T13:08:40-08:00
Nodes:
  cilium-master (localhost):
    Host connectivity to 10.0.2.15:
      ICMP:          OK, RTT=167.551µs
      HTTP via L3:   OK, RTT=780.628µs
```

Commandline options
```
$ ./cilium-health
Agent for hosting and querying the Cilium health status API

Usage:
  cilium-health [flags]
  cilium-health [command]

Available Commands:
  get         Display local cilium-health status
  ping        Check whether the cilium-health API is up
  status      Display cilium connectivity to other nodes

Flags:
      --admin string    Expose resources over 'unix' socket, 'any' socket (default "any")
  -c, --cilium string   URI to Cilium server API
  -d, --daemon          Run as a daemon
  -D, --debug           Enable debug messages
  -H, --host string     URI to cilum-health server API
  -i, --interval int    Interval (in seconds) for periodic connectivity probes (default 60)
  -p, --passive         Only respond to HTTP health checks

Use "cilium-health [command] --help" for more information about a command.
```

Next Tasks:
* [x] Launch from Cilium, similar to monitor server
* [x] Serve over TCP
  *  [x] Provide separate API path for connectivity checks
* [x] Prettify commandline output (non-JSON output)
* [x] Use RFC3339 formatting for timestamp
* [x] Extend `cilium status` to print the new fields exposed over the Cilium API
* [x] Provide access to raw json on commandline via "--json"
* [x] Provide current node name in output if available
* [x] Split IPs in API into primary / secondary
* [x] Add `--passive` option for only responding to probes
* [x] Force delete unix socket path on startup, similar to monitor
* [x] Add `--admin` option to toggle serving all APIs / unix socket

The scope of this PR is primarily to establish groundwork of determining connectivity to other nodes. Endpoint connectivity and policy connectivity is likely to be addressed in a followup PR.

For more future tasks, see Issue #1947.

joestringer committed:
daemon: Refactor getting local node addressing

This should live in pkg/node.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
api: Export cluster nodes in /status

Export a list of known cluster nodes, including names, IPs and CIDR
allocations in the health status requested on a local node.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
vendor: Add github.com/c9s/goprocinfo/linux

goprocinfo provides a wrapper around /proc on linux systems, which will
be used to collect system load in an upcoming commit.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
vendor: Add github.com/servak/go-fastping

Also add the transitive dependencies on golang.org ICMP/IPv4/IPv6.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
api: Add cilium-health API

GET /status:
  Returns the connectivity status to all other cilium-health instances
  using interval-based probing.
GET /status/probe:
  Runs a synchronous probe to all other cilium-health instances and
  returns the connectivity status.
GET /healthz:
  Returns current health:
  * uptime
  * system load
  * local agent status

See also Issue #1947

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
api: Add swagger-generated health API

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
cilium-health: Add new health status agent

cilium-health is a new client and daemon which allows users to probe the
status of the local node, and also of the connectivity between the local
node and remote nodes in the cluster.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
cilium-health: Report ping status to other nodes

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
client: Refactor formatting Cilium StatusResponse

This makes it reusable from cilium-health client.

Signed-off-by: Joe Stringer <joe@covalent.io>

aanm reviewed (commented):
Looks good so far!

joestringer committed:
client: Fix connecting to TCP servers

Previously, the client package created the client by passing the entire
"unix://foo" or "http://foo" host string into the hostname field of the
client, however when creating a client using an "http://" host string
this would result in HTTP requests attempting to reach
"http://http://hostname".

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
cilium-health: Add --passive mode for HTTP server

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
cmd: Add "--json" option.

This is only used by some commands right now, so leave it hidden (since
it's a bit inconsistent). When each command implements json-formatting
the responses, we can make this not hidden.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
cilium-health: Add support for HTTP ping to nodes

Add support to the cilium-health client and server to host a new /hello
API call over TCP socket 4240, and be able to make requests to it.

Pinging this port on all known cluster nodes is done periodically, just
like how regular ICMP ping polling is done.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
pkg: Refactor monitor launcher common code

A generic launcher will be useful for an upcoming commit where the
Cilium daemon also launches the cilium-health daemon.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
daemon: Launch cilium-health on startup

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer commented on pkg/health/server/server.go:
```diff
@@ -0,0 +1,306 @@
+
+	resp, err := s.Daemon.GetHealthz(nil)
+	if err != nil {
+		log.WithError(err).Warn("Failed to retrieve Cilium /healthz")
```
In a k8s deployment, `cilium` gets deployed, which in turn starts up `cilium-health`, but `cilium` does not serve up its API before this query goes through, so we end up with this warning in the logs because `cilium-health` started up too early. We should perhaps look at deferring the `cilium-health` launch until after the API is served (See also #2116).

joestringer replied:
Fixed.

raybejjani commented on cilium/cmd/status.go:
```diff
@@ -43,46 +44,19 @@ func statusDaemon(cmd *cobra.Command, args []string) {
+		if err != nil {
+			Fatalf("Cannot marshal status: %s", err.Error())
+		}
+		w := tabwriter.NewWriter(os.Stdout, 2, 0, 3, ' ', 0)
```
Since you already pretty printed the json, I think you can print directly to os.Stdout. In fact, the tabwriter might do funky things if there was a tab in the output.

joestringer replied:
Good point, I'll fix this up when I rebase.

raybejjani commented on pkg/client/client.go:
```diff
@@ -85,7 +88,7 @@ func NewClient(host string) (*Client, error) {
 	transport := configureTransport(nil, tmp[0], host)
 	httpClient := &http.Client{Transport: transport}
-	clientTrans := runtime_client.NewWithClient(host, clientapi.DefaultBasePath,
+	clientTrans := runtime_client.NewWithClient(tmp[1], clientapi.DefaultBasePath,
```
Why did we stop using `host` here?

joestringer replied:
The commit d955feed881d9cc044e981c687e75672347e7db5 explains it, basically the URLs for HTTP requests would be formed as `http://http://...`. This hasn't been a problem so far, because we typically host APIs over unix sockets and connect to them via default paths.

joestringer committed:
cilium-health: Add '--admin' option to restrict resources

The new '--admin' option restricts which protocols that the /healthz and
/status resources are served over. If specified as 'unix', it will only
serve these resources over the unix socket (if not otherwise disabled by
--passive); if specified as 'any' (default), these resources will also
be served over the HTTP sockets.

Signed-off-by: Joe Stringer <joe@covalent.io>

tgraf commented on api/v1/health/openapi.yaml:
```diff
@@ -0,0 +1,178 @@
+            "$ref": "../openapi.yaml#/definitions/Error"
+  "/status/probe":
+    put:
+      summary: Get connectivity status of the Cilium cluster
```
The summary should already indicate that it is synchronous

`Run synchronous connectivity probe to determine status of Cilium cluster`

joestringer replied:
Ack, will fix.

tgraf commented on api/v1/health/openapi.yaml:
```diff
@@ -0,0 +1,178 @@
+    properties:
+      cilium:
+        description: Status of Cilium daemon
+        "$ref": "../openapi.yaml#/definitions/StatusResponse"
```
So we have a `StatusResponse` in both yaml files? Can we rename one of them?

joestringer replied:
Sure, renamed the health status response.

ianvernon commented on pkg/health/server/prober.go:
```diff
@@ -0,0 +1,254 @@
+	}
+}
+
+// Run sends a single probes out to all of the other cilium nodes to gather
```
... a single probe...

cilium --> Cilium

ianvernon commented on pkg/launcher/launcher.go:
```diff
@@ -0,0 +1,126 @@
+	launcher.setStdout(stdout)
+}
+
+// Restart stops the daemon whilauncher will trigger a rerun.
```
fix typo "whilauncher"

joestringer replied:
Thanks for looking it over. I tried to split it up into reasonable separate patches for better review, but it is still quite large. I'll fix this up.

ianvernon reviewed (approved):
Looks great overall - the size made it hard to fully grok some parts, but that is the nature of bootstrapping new components :) 

Some small nits. You don't have to request another review from me.
